processor/src/trace/virtual_bus/circuit/mod.rs

```diff
@@ -0,0 +1,290 @@
+        let numerator = self.numerator * other.denominator + other.numerator * self.denominator;
+        let denominator = self.denominator * other.denominator;
+
+        Self::new(numerator, denominator)
+    }
+}
+
+/// Converts a main trace row (or more generally "query") to numerators and denominators of the
+/// input layer.
+fn evaluate_fractions_at_main_trace_query<E>(
+    query: &[E],
+    log_up_randomness: &[E],
+) -> [[E; NUM_WIRES_PER_TRACE_ROW]; 2]
+where
+    E: FieldElement,
+{
+    // numerators
+    let multiplicity = query[M_COL_IDX];
+    let f_m = {
+        let mem_selec0 = query[CHIPLETS_OFFSET];
+        let mem_selec1 = query[CHIPLETS_OFFSET + 1];
+        let mem_selec2 = query[CHIPLETS_OFFSET + 2];
+        mem_selec0 * mem_selec1 * (E::ONE - mem_selec2)
+    };
+
+    let f_rc = {
+        let op_bit_4 = query[DECODER_OP_BITS_OFFSET + 4];
+        let op_bit_5 = query[DECODER_OP_BITS_OFFSET + 5];
+        let op_bit_6 = query[DECODER_OP_BITS_OFFSET + 6];
+
+        (E::ONE - op_bit_4) * (E::ONE - op_bit_5) * op_bit_6
+    };
+
+    // denominators
+    let alphas = log_up_randomness;
+
+    let table_denom = alphas[0] - query[V_COL_IDX];
+    let memory_denom_0 = -(alphas[0] - query[MEMORY_D0_COL_IDX]);
+    let memory_denom_1 = -(alphas[0] - query[MEMORY_D1_COL_IDX]);
+    let stack_value_denom_0 = -(alphas[0] - query[DECODER_USER_OP_HELPERS_OFFSET]);
+    let stack_value_denom_1 = -(alphas[0] - query[DECODER_USER_OP_HELPERS_OFFSET + 1]);
+    let stack_value_denom_2 = -(alphas[0] - query[DECODER_USER_OP_HELPERS_OFFSET + 2]);
+    let stack_value_denom_3 = -(alphas[0] - query[DECODER_USER_OP_HELPERS_OFFSET + 3]);
+
+    [
+        [multiplicity, f_m, f_m, f_rc, f_rc, f_rc, f_rc, E::ZERO],
+        [
+            table_denom,
+            memory_denom_0,
+            memory_denom_1,
+            stack_value_denom_0,
+            stack_value_denom_1,
+            stack_value_denom_2,
+            stack_value_denom_3,
+            E::ONE,
+        ],
+    ]
+}
+
+/// Computes the wires added to the input layer that come from a given main trace row (or more
+/// generally, "query").
+fn compute_input_layer_wires_at_main_trace_query<E>(
+    query: &[E],
+    log_up_randomness: &[E],
+) -> [CircuitWire<E>; NUM_WIRES_PER_TRACE_ROW]
+where
+    E: FieldElement,
+{
+    let [numerators, denominators] =
+        evaluate_fractions_at_main_trace_query(query, log_up_randomness);
+    let input_gates_values: Vec<CircuitWire<E>> = numerators
+        .into_iter()
+        .zip(denominators)
+        .map(|(numerator, denominator)| CircuitWire::new(numerator, denominator))
+        .collect();
+    input_gates_values.try_into().unwrap()
+}
+
+/// A GKR proof for the correct evaluation of the sum of fractions circuit.
+#[derive(Debug)]
+pub struct GkrCircuitProof<E: FieldElement> {
+    circuit_outputs: CircuitLayerPolys<E>,
+    before_final_layer_proofs: BeforeFinalLayerProof<E>,
+    final_layer_proof: FinalLayerProof<E>,
+}
```

Comment:
Question: how big can we expect the proof to be (e.g., in KB) and what are the primary drivers? I'm guessing the primary driver is the complexity of the bus constraints - but does trace length also make a big difference?

Reply:
The proof is composed of two parts:

1. The round polynomials: The number of these is dependent on the trace length as well as the complexity of the bus. The degree of these is dependent on the max selector degree.
2. The opening points: The number of this dependent on the trace length and the complexity of the bus.

To give a concrete formula, suppose that the trace length is $2^{\nu}$ and the number of fractions defining the bus is $2^{\mu + 1}$, then:

1. The number of field elements defining the round polynomials is given by $$\frac{3\cdot(\mu + \nu - 1)\cdot(\mu + \nu)}{2} + 3\cdot\mu + d\cdot\nu,$$ the $3$ represents the degree of the round polynomials in the sum-checks appearing in all but the final sum-check, while the $d$ represent the degree of the round polynomial for the final sum-check (which depends as stated above on the degree of the selectors).
2. The number of opening points is given by $$4\cdot (\mu + \nu - 1) + k,$$ where $k$ is the number of trace columns, as well as their shifts, appearing in the bus.

Note that all elements in the above are extension field elements.

To give some concrete numbers, suppose that $\nu = 20$, $\mu = 5$ and $k =  5$ (max selector degree is 3), then:

1. The proof size due to round polynomials is: `1015` extension field elements which translates in the case of a quadratic extension degree to about `16Kb`.
2. The proof size due to opening points is: `136`  extension field elements  assuming we open 40 columns and column-shifts. This translates  in the case of a quadratic extension degree to about `2Kb`.